packages/sitecore-jss-react/src/components/BYOCRenderer.tsx

```diff
@@ -0,0 +1,112 @@
+  /**
+   * Name of the component to render
+   */
+  ComponentName: string;
+  /**
+   * JSON props to pass into rendered component
+   */
+  ComponentProps?: string;
+  styles?: string;
+  RenderingIdentifier?: string;
+};
+
+/**
+ * Props for BYOC wrapper component
+ */
+export type BYOCProps = {
+  /**
+   * rendering params
+   */
+  params?: BYOCRenderingParams;
+  /**
+   * fields from datasource items to be passed as rendered child component props
+   */
+  fields?: ComponentFields;
+};
+
+/**
+ * Props for BYOCRenderer component. Includes components list to load external components from.
+ */
+type BYOCRendererProps = BYOCProps & {
+  components: RegisteredComponents;
+};
+
+/**
+ * BYOCRenderer helps rendering BYOC components - that can be taken from anywhere
+ * and registered without being deployed as Sitecore renderings
+ * @param {ByocRendererProps} props component props
+ * @returns dynamicly rendered component or Missing Component error frame
+ */
+export class BYOCRenderer extends React.Component<BYOCRendererProps> {
+  state: Readonly<{ error?: Error }>;
+
+  constructor(props: BYOCRendererProps) {
+    super(props);
+    this.state = {};
+  }
+
+  static getDerivedStateFromError(error: Error) {
+    // Update state so the next render will show the fallback UI.
+    return { error: error };
+  }
+
+  componentDidCatch(error: Error) {
+    this.setState({ error });
+  }
+
+  render() {
+    const props: BYOCRendererProps = this.props;
+    if (this.state.error) {
+      return <div>A rendering error occurred: {this.state.error.message}.</div>;
+    }
+    const { ComponentName: componentName } = props.params || {};
+    if (!componentName) {
+      const noNameProps = {
+        errorOverride: 'BYOC: The ComponentName for this rendering is missing',
+      };
+      return <MissingComponent {...noNameProps} />;
+    }
+    // props.components would contain component from internal FEAAS regsitered component collection (registered in app)
+    // we can't access this collection here directly, as the collection from packages's dependency would be different from the one in app
+    const Component = props.components[componentName]?.component;
+
+    if (!Component) {
+      console.warn(
+        `Component "${componentName}" was not registered, please ensure the FEEAS.External.registerComponent call is made.`
+      );
+      const missingProps = {
+        rendering: {
+          componentName: componentName,
+        },
+        errorOverride: 'BYOC: This component was not registered.',
+      };
+      return <MissingComponent {...missingProps} />;
+    }
+
+    let componentProps: { [key: string]: unknown } = undefined;
+
+    if (props.params?.ComponentProps) {
+      try {
+        componentProps = JSON.parse(props.params.ComponentProps) ?? {};
+      } catch (e) {
+        console.warn(
+          `Parsing props for ${componentName} component from rendering params failed. Attempting to parse from data source`
+        );
```

Comment:
Doesn't look like we provide the actual error (`e`) in the log? And actually, I would consider this an error state and handle it as such (don't attempt to parse from data source).

Reply:
I disagree, ComponentProps param and data source fields are both valid sources for props. If the preferred one (ComponentProps) fails, there should be a fallback. We also don't want to throw error here and crash the whole page for something that can be fixed by editors.

Reply:
They are both valid sources, but are they valid for the same component? My thinking with that comment is based on my understanding that a component is using _one or the other, but not both_. i.e. If `params.ComponentProps` are present, the BYOC is configured to use them and those should be the authority (parsing error or not).

Also, we definitely don't want to crash the whole page - if not caught by `componentDidCatch`, then we'd return an error message directly (or `errorComponent` - see other comment).

Also, pretty sure there a number of [possible JSON.parse error messages](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/JSON_bad_parse#message).

Reply:
> Also, pretty sure there a number of [possible JSON.parse error messages](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/JSON_bad_parse#message).

this is fair, then we do need to show it.